apple-codesign/src/windows.rs

```diff
@@ -0,0 +1,736 @@
+            store_type_as_wstr.unwrap().as_ptr() as _,
+        )
+    };
+    if store_handle.is_null() {
+        return Err(AppleCodesignError::WindowsStoreError(format!("could not open store {} with store type {} (this should not happen): 0x{:08X}", store_name_as_str, store_type_as_str, get_last_error())));
+    }
+
+    // Enumerate the certificates.
+    let mut cert_context = ptr::null_mut();
+    loop {
+        // Get the next certificate.
+        cert_context = unsafe {
+            CertFindCertificateInStore(
+                store_handle,
+                X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+                0,
+                CERT_FIND_ANY,
+                ptr::null_mut(),
+                cert_context,
+            )
+        };
+        if cert_context.is_null() {
+            break;
+        }
+        let cert = match StoreCertificate::new(cert_context) {
+            Ok(cert) => cert,
+            Err(()) => continue,
+        };
+        certs.push(cert);
+    }
+
+    // Close the store.
+    unsafe {
+        CertCloseStore(store_handle, 0)
+    };
+
+    Ok(certs)
+}
+
+/// Locate code signing certificates in the Windows store.
+pub fn windows_store_find_code_signing_certificates(
+    store_name: StoreName,
+    store_type: StoreType,
+) -> Result<Vec<StoreCertificate>, AppleCodesignError> {
+    let certs = find_certificates(store_name, store_type)?;
+
+    Ok(certs
+        .into_iter()
+        .filter(|cert| !cert.captured.apple_code_signing_extensions().is_empty())
+        .collect::<Vec<_>>())
+}
+
+/// Find the x509 certificate chain for a certificate given search parameters.
+///
+/// `store_name` specifies which store to operate on.
+///
+/// `thumbprint` specifies the SHA1 thumbprint of the certificate to search for.
+/// You can find this in `certmgr.msc` or `certlm.msc` by clicking on the certificate in
+/// question and looking for `Thumbprint` under the `Details` tab.
+pub fn windows_store_find_certificate_chain(
+    store_name: StoreName,
+    thumbprint: &str,
+) -> Result<Vec<CapturedX509Certificate>, AppleCodesignError> {
+    // We look for the code signing certificate in the MY store.
+    let user_certs = find_certificates(store_name, StoreType::MY)?;
+
+    // Now search for the requested start certificate and pull the thread until
+    // we get to a self-signed certificate.
+    let start_cert: &CapturedX509Certificate = user_certs
+        .iter()
+        .find_map(|cert| {
+            if let Ok(digest) = cert
+                .captured
+                .sha1_fingerprint()
+            {
+                // Convert the Digest into a byte array
+                let digest_bytes = digest.as_ref();
+
+                // Format each byte as a two-character hex string
+                let digest_hex: String = digest_bytes
+                    .iter()
+                    .map(|byte| format!("{:02x}", byte))
+                    .collect::<String>()
+                    .to_lowercase();
```

Comment:
You can do this as a one-liner using `hex::encode()`.

Reply:
Thanks for the suggestion, should be fixed now.